pkg/causality/worker.go

```diff
@@ -24,6 +31,99 @@ type txnEvent interface {
 	GenSortedDedupKeysHash(numSlots uint64) []uint64
 }
 
-type worker[Txn txnEvent] interface {
-	Add(txn Txn, unlock func())
+// TxnWithNotifier is a wrapper of txnEvent with a PostTxnExecuted.
+type TxnWithNotifier[Txn txnEvent] struct {
+	TxnEvent Txn
+	// The PostTxnExecuted will remove the txn related Node in the conflict detector's
+	// dependency graph and resolve related dependencies for these transacitons
+	// which depend on this executed txn.
+	//
+	// NOTE: the PostTxnExecuted() must be called after the txn executed.
+	PostTxnExecuted func()
+}
+
+// WorkerOption is the option for creating a worker.
+type WorkerOption struct {
+	WorkerCount int
+	// CacheSize controls the max number of txns a worker can hold.
+	CacheSize int
+	// IsBlock indicates whether the worker should block when the cache is full.
+	IsBlock bool
+}
+
+// In current implementation, the conflict detector will push txn to the workerCache.
+type workerCache[Txn txnEvent] interface {
+	// add adds a event to the workerCache.
+	add(txn TxnWithNotifier[Txn]) bool
+	// out returns a channel to receive events which are ready to be executed.
+	out() <-chan TxnWithNotifier[Txn]
+}
+
+func newWorker[Txn txnEvent](opt WorkerOption) workerCache[Txn] {
+	log.Info("create new worker cache in conflict detector",
+		zap.Int("workerCount", opt.WorkerCount),
+		zap.Int("cacheSize", opt.CacheSize), zap.Bool("isBlock", opt.IsBlock))
+	if opt.CacheSize <= 0 {
+		log.Panic("WorkerOption.CacheSize should be greater than 0, please report a bug")
+	}
+
+	if opt.IsBlock {
+		return &boundedWorkerCacheWithBlock[Txn]{ch: make(chan TxnWithNotifier[Txn], opt.CacheSize)}
+	}
+	return nil
+	// return &boundedWorkerCache[Txn]{ch: make(chan TxnWithNotifier[Txn], opt.CacheSize)}
+}
+
+// boundedWorkerCache is a worker which has a limit on the number of txns it can hold.
+//
+//nolint:unused
+type boundedWorkerCache[Txn txnEvent] struct {
+	ch chan TxnWithNotifier[Txn]
 }
+
+//nolint:unused
+func (w *boundedWorkerCache[Txn]) add(txn TxnWithNotifier[Txn]) bool {
+	select {
+	case w.ch <- txn:
+		return true
+	default:
+		return false
+	}
+}
+
+//nolint:unused
```

Comment:
Why should we retain an unused struct?

Reply:
The golang ling tool has false positives in generic scenarios and I tried to use a new version of the tool and couldn't fix the problem. Therefore, it can only be bypassed through comments.